### PR TITLE
KAFKA-7326: KStream.print() should flush on each line for PrintStream

### PR DESCRIPTION
**Pull request by huxihx**
https://issues.apache.org/jira/browse/KAFKA-7326

PrintForeachAction should flush for PrintStream after printing the line.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by huxihx** [comment]
@mjsax Please review this minor change. Thanks.

**Review by kamalcph — COMMENTED** [review]
LGTM.

nit: On close, no need to call flush again.

**Comment by huxihx** [comment]
retest it please

**Review by mjsax — APPROVED** [review]
LGTM.

Call for second review @guozhangwang @bbejeck @vvcephei

**Review by bbejeck — APPROVED** [review]
Thanks for the patch @huxihx LGTM

**Comment by bbejeck** [comment]
failure unrelated

retest this please

**Comment by guozhangwang** [comment]
@huxihx As described in the JIRA, could you also complete the javadoc part:

```
This would include:

1. set autoFlush in the constructor of printWriter.
2. document in java-docs of KStream.print that this is for debug / testing purposes only, and it will try to flush on each record print, and hence should not be used for production usage if performance requirement is key.
```

**Comment by huxihx** [comment]
@guozhangwang  `autoFlush`  is disabled by default. Do we need to explicitly set it?

**Comment by guozhangwang** [comment]
I think either setting `autoFlush` or explicitly call `flush` (your PR) is fine for 1) above, we only need to do one of them.

**Comment by huxihx** [comment]
@guozhangwang please review again. Thanks.

**Review by guozhangwang — COMMENTED** [review]
@huxihx could you update the javadoc in `KStream.java` on top of the `print` function?

**Comment by huxihx** [comment]
Does javadoc of `KStream#peek` have to be updated also?

**Comment by guozhangwang** [comment]
> Does javadoc of KStream#peek have to be updated also?

Peek is not necessarily designed for debugging only, users may call it for production code as well. So we do not need to update its javadoc.

**Review by guozhangwang — APPROVED** [review]

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/PrintForeachAction.java [hunk 1]
```diff
@@ -51,6 +51,9 @@
     public void apply(final K key, final V value) {
         final String data = String.format("[%s]: %s", label, mapper.apply(key, value));
         printWriter.println(data);
+        if (!closable) {
```
Comment (biskinis) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/PrintForeachAction.java:54]: So then this won't work as intended for `System.out` and `System.err`...

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/PrintForeachAction.java:54]: Can you elaborate? If `printWriter == System.out || printWriter == System.err` than `printWriter.flush()` will be called each time (ie, `closeable == false` -- cf line 45 above)

Comment (biskinis) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/PrintForeachAction.java:54]: Sorry.  You are right.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/PrintForeachAction.java:54]: No need to be sorry. Glad we could clarify, and that the code is correct and we did not miss something :)

**Pull request merged**